src/espidf.rs

```diff
@@ -102,6 +102,206 @@ impl Tools {
+
+        if let Some(capture) = regex.captures(&String::from_utf8_lossy(&output.stdout)) {
+            if let Some(var) = capture.get(0) {
+                log::debug!("Match: {:?}, Version: {:?}", &var.as_str(), &self.versions);
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// get the absolute PATH
+    fn abs_export_path(&self) -> PathBuf {
+        self.install_dir.join(self.export_path.as_path())
+    }
+
+    /// Creates a Command that will echo back the current version of the tool
+    ///
+    /// Since Command is non clonable this helper is provided
+    fn test_command(&self) -> Command {
+        let cmd_abs_path = self
+            .abs_export_path()
+            .join(self.version_cmd_args[0].clone());
+
+        let mut version_cmd = std::process::Command::new(cmd_abs_path);
+        version_cmd.args(self.version_cmd_args[1..].iter().cloned());
+        version_cmd
+    }
+}
+
+/// Parsing a provided tools.json file, and return a Vec<Tool> representing a Tool version of the wanted tools
+fn parse_into_tools(
+    tools_wanted: Vec<&str>,
+    tools_json_file: PathBuf,
+    install_dir: PathBuf,
+) -> anyhow::Result<Vec<Tool>> {
+    let mut tools: Vec<Tool> = Vec::new();
+
+    let os_key = get_os_target_key().unwrap();
+
+    let mut tools_string = String::new();
+    let mut tools_file = std::fs::File::open(tools_json_file)?;
+
+    tools_file.read_to_string(&mut tools_string)?;
+
+    let parsed_file = serde_json::from_str::<serde_json::Value>(&tools_string)?;
+    let tools_object = parsed_file["tools"]
+        .as_array()
+        .expect("JSON-PARSING-ERROR: make sure the provided tools.json in the esp-idf repository is not malformed");
+
+    for tool_object in tools_object.iter().filter(|parsed_tool| {
+        tools_wanted.contains(
+            &parsed_tool["name"]
+                .as_str()
+                .expect("JSON-PARSING-ERROR: make sure the provided tools.json in the esp-idf repository is not malformed"),
+        )
+    }) {
+        let name = tool_object["name"].as_str().unwrap();
+
+        log::debug!("============================================================================");
+        log::debug!("Tool name: {name}");
+
+        // notice that export_paths inside the tools.json has the structure of "key: [ [ "path1", "path2", ... ] ,]"
+        // -> to layers of indirection to get the actual path
+        // it seams only the first array is ever used
+        let export_path = tool_object["export_paths"][0].clone();
+        let export_path: Vec<&str> = export_path
+            .as_array()
+            .and_then(|path_array| path_array.iter().map(|path| path.as_str()).collect())
+            .unwrap_or_else(Vec::new);
+
+        log::debug!("export_path: {export_path:?}");
+
+        let version_cmd_args = tool_object["version_cmd"]
+            .as_array()
+            .unwrap()
```

Comment:
These `unwrap`s here and below are precisely because the `tool_object` is untyped.